### ISSUE TITLE
urem UB propagation issues?

### ISSUE THREAD

**Issue by LebedevRI** [AliveToolkit/alive2#123]
With #122 "resolved" (by checking-out previous commit), i'm seeing the following:
```
Name: D68360 urem expand
  %r = urem i8 %x, %C
  ret i8 %r
=>
  %C_wide = zext i8 %C to i16
  %C_Magic_tmp = udiv i16 -1, %C_wide
  %C_Magic = add i16 %C_Magic_tmp, 1
  %x_wide = zext i8 %x to i16
  %Lowbits = mul i16 %x_wide, %C_Magic
  %Lowbits_wider = zext i16 %Lowbits to i32
  %C_wider = zext i8 %C to i32
  %Result_full = mul i32 %Lowbits_wider, %C_wider
  %Result_high = lshr i32 %Result_full, 16
  %r = trunc i32 %Result_high to i8
  ret i8 %r

ERROR: Value mismatch for i8 %r

Example:
i8 %x = #x12 (18)
i8 %C = #x00 (0)
i16 %C_wide = #x0000 (0)
i16 %C_Magic_tmp = #xffff (65535, -1)
i16 %C_Magic = #x0000 (0)
i16 %x_wide = #x0012 (18)
i16 %Lowbits = #x0000 (0)
i32 %Lowbits_wider = #x00000000 (0)
i32 %C_wider = #x00000000 (0)
i32 %Result_full = #x00000000 (0)
i32 %Result_high = #x00000000 (0)
Source value: #x12 (18)
Target value: #x00 (0)
```
But `i8 %C = #x00 (0)`, LHS would be undefined.
So why is it complaining?

**Comment by nunoplopes** [comment]
With latest master I get this:
```
Name: D68360 urem expand
  %r = urem i8 %x, %C
  ret i8 %r
=>
  %C_wide = zext i8 %C to i16
  %C_Magic_tmp = udiv i16 -1, %C_wide
  %C_Magic = add i16 %C_Magic_tmp, 1
  %x_wide = zext i8 %x to i16
  %Lowbits = mul i16 %x_wide, %C_Magic
  %Lowbits_wider = zext i16 %Lowbits to i32
  %C_wider = zext i8 %C to i32
  %Result_full = mul i32 %Lowbits_wider, %C_wider
  %Result_high = lshr i32 %Result_full, 16
  %r = trunc i32 %Result_high to i8
  ret i8 %r

ERROR: Value mismatch for i8 %r

Example:
i8 %x = #x02 (2)
i8 %C = undef
i16 %C_wide = #x0012 (18)
i16 %C_Magic_tmp = #x0e38 (3640)
i16 %C_Magic = #x0e39 (3641)
i16 %x_wide = #x0002 (2)
i16 %Lowbits = #x1c72 (7282)
i32 %Lowbits_wider = #x00001c72 (7282)
i32 %C_wider = #x00000010 (16)
i32 %Result_full = #x0001c720 (116512)
i32 %Result_high = #x00000001 (1)
Source value: #x02 (2)  [based on undef value]
Target value: #x01 (1)
```

The issue is that `%C` is getting `undef`. If you want this to be a constant, you should use `C1` instead, which can't be undef or poison:
```
Name: D68360 urem expand
  %r = urem i8 %x, C1
  ret i8 %r
=>
  %C_wide = zext i8 C1 to i16
  %C_Magic_tmp = udiv i16 -1, %C_wide
  %C_Magic = add i16 %C_Magic_tmp, 1
  %x_wide = zext i8 %x to i16
  %Lowbits = mul i16 %x_wide, %C_Magic
  %Lowbits_wider = zext i16 %Lowbits to i32
  %C_wider = zext i8 C1 to i32
  %Result_full = mul i32 %Lowbits_wider, %C_wider
  %Result_high = lshr i32 %Result_full, 16
  %r = trunc i32 %Result_high to i8
  ret i8 %r
```

BTW, it's important to have a recent Z3 so you can get the counterexample above. I fixed a bug a while ago in Z3's model cache (I don't remember which release it integrated).

**Comment by LebedevRI** [comment]
> BTW, it's important to have a recent Z3 so you can get the counterexample above. I fixed a bug a while ago in Z3's model cache (I don't remember which release it integrated).

```
$ dpkg -l | grep z3
ii  libz3-4:amd64                                               4.8.4-1                                                 amd64        theorem prover from Microsoft Research - runtime libraries
ii  libz3-dev:amd64                                             4.8.4-1                                                 amd64        theorem prover from Microsoft Research - development files
ii  python-z3                                                   4.8.4-1                                                 amd64        theorem prover from Microsoft Research - Python bindings
ii  z3                                                          4.8.4-1                                                 amd64        theorem prover from Microsoft Research
```
Which is currently the newest version [available in debian](https://tracker.debian.org/pkg/z3).
If that is too old, _please_ reflect that in cmake version check :)



> The issue is that `%C` is getting `undef`. If you want this to be a constant, you should use `C1` instead, which can't be undef or poison:
> 
> ```
> Name: D68360 urem expand
>   %r = urem i8 %x, C1
>   ret i8 %r
> =>
>   %C_wide = zext i8 C1 to i16
>   %C_Magic_tmp = udiv i16 -1, %C_wide
>   %C_Magic = add i16 %C_Magic_tmp, 1
>   %x_wide = zext i8 %x to i16
>   %Lowbits = mul i16 %x_wide, %C_Magic
>   %Lowbits_wider = zext i16 %Lowbits to i32
>   %C_wider = zext i8 C1 to i32
>   %Result_full = mul i32 %Lowbits_wider, %C_wider
>   %Result_high = lshr i32 %Result_full, 16
>   %r = trunc i32 %Result_high to i8
>   ret i8 %r
> ```

That doesn't help:
```
$ /repositories/alive2/build-Clang-release/alive -smt-to:100000 /tmp/urem.opt 
Processing /tmp/urem.opt..

----------------------------------------
Name: D68360 urem expand
  %r = urem i4 %x, C1
  ret i4 %r
=>
  %C_wide = zext i4 C1 to i8
  %C_Magic_tmp = udiv i8 -1, %C_wide
  %C_Magic = add i8 %C_Magic_tmp, 1
  %x_wide = zext i4 %x to i8
  %Lowbits = mul i8 %x_wide, %C_Magic
  %Lowbits_wider = zext i8 %Lowbits to i16
  %C_wider = zext i4 C1 to i16
  %Result_full = mul i16 %Lowbits_wider, %C_wider
  %Result_high = lshr i16 %Result_full, 8
  %r = trunc i16 %Result_high to i4
  ret i4 %r

ERROR: Value mismatch for i4 %r

Example:
i4 %x = #x8 (8, -8)
i8 %C_wide = #x00 (0)
i8 %C_Magic_tmp = #xff (255, -1)
i8 %C_Magic = #x00 (0)
i8 %x_wide = #x08 (8)
i8 %Lowbits = #x00 (0)
i16 %Lowbits_wider = #x0000 (0)
i16 %C_wider = #x0000 (0)
i16 %Result_full = #x0000 (0)
i16 %Result_high = #x0000 (0)
Source value: #x8 (8, -8)
Target value: #x0 (0)
```
Works on alive1 - https://rise4fun.com/Alive/iof8
Is this some z3 version issue?

**Comment by nunoplopes** [comment]
I'm afraid it might be an issue with Z3.  4.8.4 is 1 year old; . Can you try compiling from git pls?
It's just:  `./configure && cd build && make && sudo make install`

**Comment by LebedevRI** [comment]
I'm not going to do that, although i did just rebuilt debian z3 package with z3-4.8.6
and installed it via package manager, like one should.

Did a clean rebuild of alive2, `ninja check` passes.
But still:
```
----------------------------------------
Name: D68360 urem expand
  %r = urem i4 %x, C1
  ret i4 %r
=>
  %C_wide = zext i4 C1 to i8
  %C_Magic_tmp = udiv i8 -1, %C_wide
  %C_Magic = add i8 %C_Magic_tmp, 1
  %x_wide = zext i4 %x to i8
  %Lowbits = mul i8 %x_wide, %C_Magic
  %Lowbits_wider = zext i8 %Lowbits to i16
  %C_wider = zext i4 C1 to i16
  %Result_full = mul i16 %Lowbits_wider, %C_wider
  %Result_high = lshr i16 %Result_full, 8
  %r = trunc i16 %Result_high to i4
  ret i4 %r

ERROR: Value mismatch for i4 %r

Example:
i4 %x = #x8 (8, -8)
i8 %C_wide = #x00 (0)
i8 %C_Magic_tmp = #xff (255, -1)
i8 %C_Magic = #x00 (0)
i8 %x_wide = #x08 (8)
i8 %Lowbits = #x00 (0)
i16 %Lowbits_wider = #x0000 (0)
i16 %C_wider = #x0000 (0)
i16 %Result_full = #x0000 (0)
i16 %Result_high = #x0000 (0)
Source value: #x8 (8, -8)
Target value: #x0 (0)
```

Let me submit a PR to bump the version check.
I'll file debian bug later.

**Comment by nunoplopes** [comment]
This is a dup of https://github.com/AliveToolkit/alive2/issues/64
TL;DR: when there's UB involved you need `-root-only` until that feature is implemented

**Comment by LebedevRI** [comment]
Thanks!
Sorry for the noise (:

**Comment by nunoplopes** [comment]
np. I know this is annoying; I get caught in this trap myself once in a while..
